types/src/checksummed_hex.rs

```diff
@@ -0,0 +1,258 @@
+    hasher.update(data);
+    hasher.finalize().to_vec()
+}
+
+/// If `input` is not greater than [`SMALL_BYTES_COUNT`], returns the bytes encoded as hexadecimal
+/// with mixed-case based checksums following a scheme similar to [EIP-55][1].  If `input` is
+/// greater than `SMALL_BYTES_COUNT`, no mixed-case checksumming is applied, and lowercase hex is
+/// returned.
+///
+/// Key differences:
+///   - Works on any length of data up to `SMALL_BYTES_COUNT`, not just 20-byte addresses
+///   - Uses Blake2b hashes rather than Keccak
+///   - Uses hash bits rather than nibbles
+///
+/// [1]: https://eips.ethereum.org/EIPS/eip-55
+pub fn encode(input: &(impl AsRef<[u8]> + ?Sized)) -> String {
+    if input.as_ref().len() > SMALL_BYTES_COUNT {
+        return base16::encode_lower(input);
+    }
+    encode_iter(input).collect()
+}
+
+/// `encode` but it returns an iterator.
+fn encode_iter(input: &(impl AsRef<[u8]> + ?Sized)) -> impl Iterator<Item = char> + '_ {
+    let input_bytes = input.as_ref();
+    let nibbles = bytes_to_nibbles(input_bytes);
+    let mut hash_bits = bytes_to_bits_cycle(blake2b_hash(input));
+    nibbles.map(move |mut nibble| {
+        // Base 16 numbers greater than 10 are represented by the ascii characters a through f.
+        if nibble >= 10 && hash_bits.next().unwrap_or(true) {
+            // We are using nibble to index HEX_CHARS, so adding 6 to nibble gives us the index
+            // of the uppercase character. HEX_CHARS[10] == 'a', HEX_CHARS[16] == 'A'.
+            nibble += 6;
+        }
+        HEX_CHARS[nibble as usize]
+    })
+}
+
+/// Returns true if all chars in a string are uppercase or lowercase.
+/// Returns false if the string is mixed case or if there are no alphabetic chars.
+fn string_is_same_case<T: AsRef<[u8]> + ?Sized>(s: &T) -> bool {
+    const LOWER_RANGE: RangeInclusive<u8> = b'a'..=b'f';
+    const UPPER_RANGE: RangeInclusive<u8> = b'A'..=b'F';
+
+    let mut chars = s
+        .as_ref()
+        .iter()
+        .filter(|c| LOWER_RANGE.contains(c) || UPPER_RANGE.contains(c));
+
+    match chars.next() {
+        Some(first) => {
+            let is_upper = UPPER_RANGE.contains(first);
+            chars.all(|c| UPPER_RANGE.contains(c) == is_upper)
+        }
+        None => {
+            // String has no actual characters.
+            true
+        }
+    }
+}
+
+/// Decodes a mixed-case hexadecimal string, verifying that it conforms to the checksum scheme
+/// similar to scheme in [EIP-55][1].
+///
+/// Key differences:
+///   - Works on any length of (decoded) data up to `SMALL_BYTES_COUNT`, not just 20-byte addresses
+///   - Uses Blake2b hashes rather than Keccak
+///   - Uses hash bits rather than nibbles
+///
+/// For backward compatibility: if the hex string is all uppercase or all lowercase, the check is
+/// skipped.
+///
+/// [1]: https://eips.ethereum.org/EIPS/eip-55
+pub fn decode(input: &(impl AsRef<[u8]> + ?Sized)) -> Result<Vec<u8>, base16::DecodeError> {
```

Comment:
```suggestion
pub fn decode(input: T) -> Result<Vec<u8>, base16::DecodeError>
where
    T: AsRef<[u8]> + ?Sized,
{
```

or

```rust
pub fn decode(input: impl AsRef<[u8]> + ?Sized) -> Result<Vec<u8>, base16::DecodeError> {
```

Not sure why we need a ref here...

Reply:
Fixed in [e93dc37](https://github.com/nilz3ro/casper-node/pull/5/commits/e93dc3787a9983c352ebfdba0fc6bccb8f4d7c41).